### PR TITLE
feat: add directory method, support deep directories

### PR DESCRIPTION
**Pull request by sveisvei**
## JIRA Issue
[COREWEB-83](https://jira.finn.no/browse/COREWEB-83)

## Status
**READY**

## Description
Add a new async method for listing files in a directory.

It returns an array with objects of shape: `{fileName, content}`

## Todos
- [x] Tests
- [x] Documentation

## Related PRs
* https://github.com/asset-pipe/asset-pipe-sink-mem/pull/48
* https://github.com/asset-pipe/asset-pipe-sink-gcs/pull/50

### CODE REVIEW

File: lib/sink.js [hunk 1]
```diff
@@ -5,19 +5,22 @@ const JSONStream = require('JSONStream');
 const common = require('asset-pipe-common');
 const stream = require('readable-stream');
 const assert = require('assert');
-const path = require('path');
+const pathLib = require('path');
```
Comment (digitalsadhu) [inline lib/sink.js]: you could destructure `join` out to avoid a clash.

File: lib/sink.js [hunk 2]
```diff
@@ -103,28 +106,54 @@ module.exports = class SinkFs extends EventEmitter {
...
+                this.joinPath(fileName),
                 fs.constants.R_OK | fs.constants.W_OK
             );
             return true;
         } catch (e) {}
         return false;
     }
 
+    async directory(dirName = '/') {
+        const resolveTagetDirName = this.joinPath(dirName);
```
Comment (digitalsadhu) [inline lib/sink.js]: s/resolveTaget/DirName/resolveTargetDirName

File: lib/sink.js [hunk 3]
```diff
@@ -103,28 +106,54 @@ module.exports = class SinkFs extends EventEmitter {
...
+        const dir = await readdir(resolveTagetDirName);
+        return Promise.all(
+            dir.map(relativePath =>
+                stat(pathLib.join(resolveTagetDirName, relativePath)).then(
```
Comment (digitalsadhu) [inline lib/sink.js]: typo here too.

File: lib/sink.js [hunk 4]
```diff
@@ -103,28 +106,54 @@ module.exports = class SinkFs extends EventEmitter {
...
+                        ).then(content => ({
+                            fileName: relativePath,
+                            content,
+                        }));
+                    }
+                )
+            )
+        ).then(results => results.filter(Boolean));
+    }
+
```
Comment (digitalsadhu) [inline lib/sink.js]: async/await?

File: test/sink.test.js [hunk 5]
```diff
@@ -85,6 +91,30 @@ test('.writer() - no value for "type" argument - should throw', () => {
...
+    expect(directoryContent[0].content.length).toMatchSnapshot();
+});
+
+test('directory() - should fetch all files in sub directory', async () => {
+    const sink = new Sink({ path: './test/mock/' });
+    const directoryContent = await sink.directory('/sub');
+    expect(directoryContent).toMatchSnapshot();
+});
+
+test('directory() - should error when directory is missing', async () => {
```
Comment (sveisvei) [inline test/sink.test.js]: Should it error?

File: test/sink.test.js [hunk 6]
```diff
@@ -46,6 +46,12 @@ test('.set() - should set value', async () => {
     expect(await sink.get('some-key-1')).toBe('value-1');
 });
 
+test('.set() - should set a deep folder', async () => {
+    const sink = getFreshSink('set-deep-1');
```
Comment (digitalsadhu) [inline test/sink.test.js]: now thats fresh 💨

File: test/sink.test.js [hunk 7]
```diff
@@ -85,6 +91,30 @@ test('.writer() - no value for "type" argument - should throw', () => {
...
+    expect(directoryContent).toMatchSnapshot();
+});
+
+test('directory() - should error when directory is missing', async () => {
+    expect.assertions(1);
+    const sink = new Sink({ path: './test/mock/' });
+    try {
+        await sink.directory('/missing');
+    } catch (e) {
+        expect(e.code).toMatchSnapshot();
```
Comment (digitalsadhu) [inline test/sink.test.js]: is this the same error behaviour/message as for other sinks?

File: test/sink.test.js (hunk 5 above)
Comment (digitalsadhu) [reply test/sink.test.js]: I wondered that too. I think as long as we are consistent across syncs and document the behaviour we should be good.

Comment (digitalsadhu) [reply test/sink.test.js]: perhaps lean towards erroring with clear usage fail message. Thats probably clearer then wondering what it does if you don't give it a path (as a consumer)

File: test/sink.test.js (hunk 7 above)
Comment (sveisvei) [reply test/sink.test.js]: No - not atm - but it might have to be?

Comment (digitalsadhu) [reply test/sink.test.js]: I think so. Be nice if the error mechanism and messaging were the same across sinks

Comment (sveisvei) [reply test/sink.test.js]: Will fix. I guess we just want an error for "missing folder or empty result"?

Comment (digitalsadhu) [reply test/sink.test.js]: yeah that sounds good

File: lib/sink.js [hunk 8]
```diff
@@ -58,7 +58,7 @@ class WriteStream extends stream.PassThrough {
 class ReadStream extends fs.createReadStream {
     constructor(...args) {
         super(...args);
-        const file = path.parse(args[0]).base;
+        const file = parse(args[0]).base;
```
Comment (SimenB) [inline lib/sink.js]: not `path.basename`?

File: test/sink.test.js [hunk 9]
```diff
@@ -85,6 +91,30 @@ test('.writer() - no value for "type" argument - should throw', () => {
     }).toThrowError('"type" is missing');
 });
 
+test('dir() - should fetch file in root directory', async () => {
+    const sink = new Sink({ path: './test/mock' });
+    const directoryContent = await sink.dir('/');
+    expect(directoryContent.length).toBe(1);
```
Comment (SimenB) [inline test/sink.test.js]: `toHaveLength`

File: lib/sink.js (hunk 8 above)
Comment (sveisvei) [reply lib/sink.js]: fixed.

**Pull request merged**